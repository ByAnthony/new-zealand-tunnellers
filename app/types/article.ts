// Database extract
export type SectionData = {
  title: string;
  text: string;
};

export type ImageData = {
  file: string;
  title: string | null;
  photographer: string | null;
  reference: string | null;
  alt: string;
};

export type ArticleReferenceData = {
  id: string;
  chapter: number;
  title: string;
};

export type AboutUsData = {
  id: string;
  title: string;
};

export type ArticleData = AboutUsData & {
  chapter: number;
  notes: string;
};

// Shaped data
export type Section = {
  title: string;
  text: string;
};

export type ImageArticle = {
  file: string;
  title: string | null;
  photographer: string | null;
  reference: string | null;
  alt: string;
};

export type Next = {
  url: string;
  chapter: number;
  title: string;
};

export type AboutUsArticle = {
  id: string;
  title: string;
  section: Section[];
  image: ImageArticle[];
};

export type Chapter = AboutUsArticle & {
  chapter: number;
  title: string;
  next: Next | null;
  notes: string;
};
