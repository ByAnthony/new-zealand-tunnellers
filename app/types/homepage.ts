export type TunnellerImages = {
  id: string;
  image: string;
};

export type historyChapter = {
  id: string;
  chapter: number;
  title: string;
  image: string;
};

export type historyImageChapters = {
  file: string;
};
