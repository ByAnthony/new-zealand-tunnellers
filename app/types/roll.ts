export type Name = {
    surname: string,
    forename: string,
  }
  
export type Tunneller = {
    id: number,
    name: Name,
    birthDate: string | null,
    deathDate: string | null,
  }
