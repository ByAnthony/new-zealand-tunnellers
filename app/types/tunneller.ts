import { Name } from "./tunnellers";

// Database extract
export type ProfileData = {
  id: number;
  surname: string;
  forename: string;
  birth_date: string | null;
  death_date: string | null;
  birth_country: string | null;
  mother_name: string | null;
  mother_origin: string | null;
  father_name: string | null;
  father_origin: string | null;
  nz_resident_in_month: number | null;
  enlistment_date: string | null;
  posted_date: string | null;
  occupation: string | null;
  employer: string | null;
  residence: string | null;
  marital_status: string | null;
  wife_name: string | null;
  serial: string;
  rank: string;
  district: string | null;
  aka: string | null;
  posted_from_corps: string | null;
  embarkation_unit: string;
  training_start: string;
  training_location: string;
  training_location_type: string;
  section: string | null;
  attached_corps: string | null;
  transport_uk_ref: string;
  transport_uk_vessel: string;
  transport_uk_start: string;
  has_deserted: number | null;
  transferred_to_date: string | null;
  transferred_to_unit: string | null;
  death_type: string | null;
  transport_nz_ref: string | null;
  transport_nz_vessel: string | null;
  transport_nz_start: string | null;
  demobilization_date: string | null;
  discharge_uk: number | null;
  death_location: string | null;
  death_town: string | null;
  death_country: string | null;
  death_cause: string | null;
  death_circumstances: string | null;
  cemetery: string | null;
  cemetery_town: string | null;
  cemetery_country: string | null;
  grave: string | null;
  awmm_cenotaph: string | null;
  nominal_roll_volume: string | null;
  nominal_roll_number: string | null;
  nominal_roll_page: string;
  image: string | null;
  image_source_auckland_libraries: string | null;
  archives_name: string | null;
  archives_ref: string | null;
  family_name: string | null;
  newspaper_name: string | null;
  newspaper_date: string | null;
  book_title: string | null;
  book_town: string | null;
  book_publisher: string | null;
  book_year: string | null;
  book_page: string | null;
};

export type DeathData = {
  deathType: string | null;
  deathDate: string | null;
  deathLocation: string | null;
  deathTown: string | null;
  deathCountry: string | null;
  deathCause: string | null;
  deathCircumstances: string | null;
  cemetery: string | null;
  cemteryTown: string | null;
  cemeteryCountry: string | null;
  grave: string | null;
};

export type SingleEventData = {
  date: string;
  event: string;
  title: string | null;
  image: string | null;
};

export type EventData = {
  date: DateObj;
  event: string;
  title: string;
  image: string | null;
};

export type JoinEventData = {
  enlistmentDate: string | null;
  trainingStart: string;
  trainingLocation: string;
  embarkationUnit: string;
};

// Shaped data
export type DateObj = {
  year: string;
  dayMonth: string;
};

export type Summary = {
  serial: string;
  name: Name;
  birth: string | null;
  death: string | null;
};

export type Birth = {
  date: DateObj | null;
  country: string | null;
};

export type Parent = {
  name: string;
  origin: string | null;
};

export type Parents = {
  mother: Parent | null;
  father: Parent | null;
};

export type Origins = {
  birth: Birth;
  parents: Parents;
  inNzLength: string | null;
};

export type ArmyExperience = {
  unit: string;
  country: string | null;
  conflict: string | null;
  duration: string | null;
};

export type Employment = {
  occupation: string | null;
  employer: string | null;
};

export type PreWayYears = {
  armyExperience: ArmyExperience[];
  employment: Employment;
  residence: string | null;
  maritalStatus: string | null;
  wife: string | null;
};

export type Transferred = {
  date: DateObj | null;
  unit: string | null;
};

export type Enlistment = {
  serial: string;
  rank: string;
  date: DateObj | null;
  district: string | null;
  alias: string | null;
  transferredToTunnellers: Transferred | null;
  ageAtEnlistment: number | null;
};

export type Training = {
  date: DateObj | null;
  location: string;
  locationType: string;
};

export type EmbarkationUnit = {
  detachment: string;
  training: Training;
  section: string | null;
  attachedCorps: string | null;
};

export type Transport = {
  reference: string | null;
  vessel: string;
  departureDate: DateObj;
};

export type EventDetail = {
  description: string;
  title: string | null;
  image: string | null;
};

export type Event = {
  date: DateObj;
  event: EventDetail[];
};

export type TransferredTo = {
  date: DateObj | null;
  unit: string;
};

export type Medal = {
  name: string;
  country: string;
  image: string;
  citation: string | null;
};

export type Demobilization = {
  date: DateObj;
  country: string | null;
};

export type EndOfService = {
  deserter: boolean;
  transferred: TransferredTo | null;
  deathWar: boolean;
  transportNz: Transport | null;
  demobilization: Demobilization | null;
};

export type MilitaryYears = {
  enlistment: Enlistment;
  embarkationUnit: EmbarkationUnit;
  transportUk: Transport | null;
  frontEvents: Record<string, Event[]>;
  endOfService: EndOfService;
  medals: Medal[];
};

export type DeathPlace = {
  location: string | null;
  town: string | null;
  country: string | null;
};

export type DeathCause = {
  cause: string | null;
  circumstances: string | null;
};

export type Cemetery = {
  name: string | null;
  location: string | null;
  country: string | null;
  grave: string | null;
};

export type Death = {
  warInjuriesDeathAfterWar: boolean;
  date: DateObj | null;
  place: DeathPlace | null;
  cause: DeathCause | null;
  cemetery: Cemetery | null;
  ageAtDeath: number | null;
};

export type NzArchives = {
  reference: string;
  url: string;
};

export type Book = {
  title: string;
  town: string;
  publisher: string;
};

export type NominalRoll = Book & {
  date: string;
  page: string;
  volume?: string | null;
  roll?: string | null;
};

export type LondonGazette = {
  page: string;
  date: string;
};

export type Sources = {
  nzArchives: NzArchives[];
  awmmCenotaph: string | null;
  nominalRoll: NominalRoll;
  londonGazette: LondonGazette[];
};

export type ImageArchives = {
  location: string | null;
  reference: string | null;
};

export type ImageNewspaper = {
  name: string | null;
  date: string | null;
};

export type Author = {
  forename: string;
  surname: string;
};

export type ImageBook = {
  authors: Author[] | null;
  title: string | null;
  town: string | null;
  publisher: string | null;
  year: string | null;
  page: string | null;
};

export type ImageSource = {
  aucklandLibraries: string | null;
  archives: ImageArchives | null;
  family: string | null;
  newspaper: ImageNewspaper | null;
  book: ImageBook | null;
};

export type ImageTunneller = {
  url: string;
  source: ImageSource;
};

export type TunnellerProfile = {
  id: number;
  summary: Summary;
  origins: Origins;
  preWarYears: PreWayYears;
  militaryYears: MilitaryYears;
  death: Death | null;
  sources: Sources;
  image: ImageTunneller | null;
};
