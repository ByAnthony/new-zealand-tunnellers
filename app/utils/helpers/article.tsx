import {
  ArticleReferenceData,
  ImageData,
  SectionData,
} from "../../types/article";

export const getSections = (sections: SectionData[]) => {
  return sections.map((section: SectionData) => ({
    title: section.title,
    text: section.text,
  }));
};

export const getImages = (images: ImageData[]) => {
  return images.map((image: ImageData) => ({
    file: image.file,
    title: image.title,
    photographer: image.photographer,
    reference: image.reference,
    alt: image.alt,
  }));
};

export const getNextChapter = (
  chapter: number,
  articles: ArticleReferenceData[],
) => {
  const index = articles.findIndex(
    (article: ArticleReferenceData) => article.chapter === chapter,
  );

  if (index !== -1 && index + 1 < articles.length) {
    const nextArticle = articles[index + 1];
    return {
      url: nextArticle.id,
      chapter: nextArticle.chapter,
      title: nextArticle.title,
    };
  }

  return null;
};

export const formatText = (text: string) => {
  const paragraphs = text.split("\\n\\n");

  let segmentKey: number = 0;

  return paragraphs.map((paragraph) => {
    const segments = paragraph
      .replace(/--/g, "\u00A0")
      .split(/(\*.+?\*)|(\[.+?\))/g);

    const formattedSegments = segments.map((segment) => {
      segmentKey += 1;
      if (segment && segment.startsWith("*") && segment.endsWith("*")) {
        const italicText = segment.slice(1, -1);
        return <em key={segmentKey}>{italicText}</em>;
      }
      if (segment && segment.startsWith("[") && segment.endsWith(")")) {
        const linkText = segment.slice(1, -1);

        const [label, url] = linkText.split("](");

        if (url.includes("footnote")) {
          return (
            <a key={segmentKey} href={url} id={`reference_${label}`}>
              {`[${label}]`}
            </a>
          );
        }
        if (url.includes("reference")) {
          return (
            <a
              key={segmentKey}
              href={url}
              id={`footnote_${label.slice(0, -1)}`}
            >
              {label}
            </a>
          );
        }
        return (
          <a key={segmentKey} href={url}>
            {label}
          </a>
        );
      }
      return <span key={segmentKey}>{segment}</span>;
    });

    return <p key={paragraphs.indexOf(paragraph)}>{formattedSegments}</p>;
  });
};
