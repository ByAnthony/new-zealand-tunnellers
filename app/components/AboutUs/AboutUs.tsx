"use client";

import Image from "next/image";
import { AboutUsArticle } from "../../types/article";
import { Paragraph } from "../Article/Paragraph/Paragraph";
import { Title } from "../Title/Title";

import STYLES from "./AboutUs.module.scss";

type Props = {
  article: AboutUsArticle;
};

export function AboutUs({ article }: Props) {
  return (
    <div className={STYLES.container}>
      <div className={STYLES.header}>
        <Title title={article.title} />
      </div>
      <Paragraph section={article.section[0]} />
      <div className={STYLES["image-container"]}>
        <Image
          src={`/images/about-us/${article.image[0].file}`}
          alt={article.image[0].alt}
          width={800}
          height={575}
          className={STYLES.image}
        />
      </div>
      <Paragraph section={article.section[1]} />
    </div>
  );
}
