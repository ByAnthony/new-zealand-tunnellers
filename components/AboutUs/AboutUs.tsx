"use client";

import Image from "next/image";
import { useEffect } from "react";

import { Paragraph } from "@/components/Article/Paragraph/Paragraph";
import { Title } from "@/components/Title/Title";
import { AboutUsArticle } from "@/types/article";

import STYLES from "./AboutUs.module.scss";

type Props = {
  article: AboutUsArticle;
};

export function AboutUs({ article }: Props) {
  useEffect(() => {
    window.localStorage.clear();
  }, []);

  return (
    <div className={STYLES.container}>
      <div className={STYLES.header}>
        <Title title={article.title} />
      </div>
      <Paragraph section={article.section[0]} />
      <div className={STYLES["image-container"]}>
        <Image
          src={`/images/about-us/${article.image[0].file}`}
          alt={article.image[0].alt}
          width={800}
          height={575}
          className={STYLES.image}
          placeholder="empty"
        />
      </div>
      <div className={STYLES["get-in-touch"]}>
        <Paragraph section={article.section[2]} />
        <div className={STYLES["contact-buttons"]}>
          <button
            type="button"
            className={STYLES.email}
            onClick={() => window.open("mailto:info@nztunnellers.com")}
            aria-label="Contact by email"
          >
            Email
          </button>
          <button
            type="button"
            className={STYLES.linkedin}
            onClick={() =>
              (window.location.href =
                "https://www.linkedin.com/in/anthony-byledbal/")
            }
            aria-label="Contact on LinkedIn"
          >
            LinkedIn
          </button>
        </div>
      </div>
      <Paragraph section={article.section[1]} />
    </div>
  );
}
