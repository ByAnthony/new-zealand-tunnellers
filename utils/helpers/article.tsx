import Link from "next/link";

import { ArticleReferenceData } from "@/types/article";

export const getNextChapter = (
  chapter: number,
  articles: ArticleReferenceData[],
) => {
  const index = articles.findIndex(
    (article: ArticleReferenceData) => article.chapter === chapter,
  );

  if (index !== -1 && index + 1 < articles.length) {
    const nextArticle = articles[index + 1];
    return {
      url: nextArticle.id,
      chapter: nextArticle.chapter,
      title: nextArticle.title,
    };
  }

  return null;
};

export const formatText = (text: string) => {
  const paragraphs = text.split("\\n\\n");

  let segmentKey: number = 0;

  return paragraphs.map((paragraph) => {
    const segments = paragraph
      .replace(/--/g, "\u00A0")
      .split(/(\*.+?\*)|(\[.+?\))/g);

    const formattedSegments = segments
      .filter((segment) => segment)
      .map((segment) => {
        segmentKey += 1;

        if (segment && segment.startsWith("*") && segment.endsWith("*")) {
          const italicText = segment.slice(1, -1);
          return <em key={segmentKey}>{italicText}</em>;
        }

        if (segment && segment.startsWith("[") && segment.endsWith(")")) {
          const linkText = segment.slice(1, -1);

          const [label, url] = linkText.split("](");

          if (url.includes("footnote")) {
            return (
              <a
                key={segmentKey}
                href={url}
                id={`reference_${label}`}
                className="anchor-link"
                aria-label={`Jump to reference ${label} for the source of information`}
              >
                {`[${label}]`}
              </a>
            );
          }
          if (url.includes("reference")) {
            return (
              <a
                key={segmentKey}
                href={url}
                id={`footnote_${label.slice(0, -1)}`}
                className="anchor-link"
                aria-label={`Jump back to the footnote ${label.slice(0, -1)}`}
              >
                {label}
              </a>
            );
          }
          return (
            <Link key={segmentKey} href={url}>
              {label}
            </Link>
          );
        }

        return <span key={segmentKey}>{segment}</span>;
      });

    return <p key={paragraphs.indexOf(paragraph)}>{formattedSegments}</p>;
  });
};
