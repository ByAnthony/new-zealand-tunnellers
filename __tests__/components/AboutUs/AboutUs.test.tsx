import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { AboutUs } from "../../../app/components/AboutUs/AboutUs";
import { AboutUsArticle } from "../../../app/types/article";

describe("<AboutUs />", () => {
  const mockArticle: AboutUsArticle = {
    id: "about-us",
    title: "Our Story",
    section: [
      { title: "Title 1", text: "This is the first paragraph of our story." },
      { title: "Title 2", text: "This is the second paragraph of our story." },
    ],
    image: [
      {
        file: "our-story.jpg",
        title: null,
        photographer: null,
        reference: null,
        alt: "A picture of our team",
      },
    ],
  };

  test("matches the snapshot", () => {
    const { asFragment } = render(<AboutUs article={mockArticle} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders the article title, sections, and image correctly", () => {
    render(<AboutUs article={mockArticle} />);

    expect(screen.getByText("Our Story")).toBeInTheDocument();

    expect(
      screen.getByText("This is the first paragraph of our story."),
    ).toBeInTheDocument();
    expect(
      screen.getByText("This is the second paragraph of our story."),
    ).toBeInTheDocument();

    const image = screen.getByRole("img");
    expect(image).toHaveAttribute("alt", "A picture of our team");
  });
});
