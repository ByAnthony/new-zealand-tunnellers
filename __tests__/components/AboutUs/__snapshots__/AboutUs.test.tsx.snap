// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AboutUs /> matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="main-title"
      >
        <h1>
          <span
            class="title-line-1"
          >
            Our Story
          </span>
          <span
            class="title-line-2"
          />
        </h1>
      </div>
    </div>
    <div
      class="paragraph"
    >
      <h2>
        Title 1
      </h2>
      <p>
        <span>
          This is the first paragraph of our story.
        </span>
      </p>
    </div>
    <div
      class="image-container"
    >
      <img
        alt="A picture of our team"
        class="image"
        data-nimg="1"
        decoding="async"
        height="575"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Fabout-us%2Four-story.jpg&w=1920&q=75"
        srcset="/_next/image?url=%2Fimages%2Fabout-us%2Four-story.jpg&w=828&q=75 1x, /_next/image?url=%2Fimages%2Fabout-us%2Four-story.jpg&w=1920&q=75 2x"
        style="color: transparent;"
        width="800"
      />
    </div>
    <div
      class="paragraph"
    >
      <h2>
        Title 2
      </h2>
      <p>
        <span>
          This is the second paragraph of our story.
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
